### PR TITLE
Optimize baker_hubbard and wernet_nilsson functions

### PR DESCRIPTION
**Pull request by connorbrinton**
Hi MDTraj team,

Here are some changes I've made to the hbond-finding functions to make them run about six times faster, as well as adding the option sidechain_only, to consider only sidechain-sidechain interactions. I still have to write a test for the code, but in my ad-hoc testing all outputs have been identical to previous versions of mdtraj, just faster. Feel free to integrate this code however you like.

Note: I also haven't added documentation for the sidechain_only option, which should be done if it's something you want to support.

Thanks,
Connor

### CODE REVIEW

**Comment by connorbrinton** [comment]
Also, just so you know what to look for, the following were the optimizations I made:
(1) Generate donor-hydrogen-acceptor lists using numpy (instead of using itertools.product)
(2) Calculate angles using distances (from compute_distances, which is really fast) and the law of cosines (rather than using compute_angles)
(3) Don't calculate angles for bonds that don't satisfy the distance criteria (this one is the biggest improvement)

File: mdtraj/geometry/hbond.py [hunk 1]
```diff
@@ -38,8 +37,7 @@
 # Functions
 ##############################################################################
 
-
-def wernet_nilsson(traj, exclude_water=True, periodic=True):
+def wernet_nilsson(traj, exclude_water=True, sidechain_only=False, periodic=True):
```
Comment (mpharrigan) [inline mdtraj/geometry/hbond.py]: Can you add `sidechain_only` to the end of this list. This makes sure it maintains API compatibility when using positional arguments

**Comment by mpharrigan** [comment]
Looks like this also does some badly-needed refactoring cc #705

**Comment by mpharrigan** [comment]
Maybe also related #522

**Comment by connorbrinton** [comment]
I moved the sidechain_only option, let me know if there's anything else that should be changed.

**Comment by connorbrinton** [comment]
Looks like there are tests for hbonds, and that some of them didn't pass. I'll figure out why, and fix the code if necessary.

**Comment by connorbrinton** [comment]
So the failed tests were definitely my fault (oops), but I found the things I had done wrong and fixed them, so the tests are passing now. Hooray!

**Comment by connorbrinton** [comment]
I spoke too soon last Friday saying that all the tests were passing, because some of the 64-bit builds hadn't finished yet. They gave different inputs (empty trajectories) to the hbond functions, so they exposed problems with empty trajectories, which I fixed this morning.

**Comment by connorbrinton** [comment]
@mpharrigan could you check the AppVeyor and Travis CI builds to see if the two failed builds are because of my changes?  From what I could tell from the console the AppVeyor build failure was because of a build system problem and the Travis CI build failure was because of a race condition in a SASA test. But if they're because of something I did, I'll fix it. Thanks!

**Comment by mpharrigan** [comment]
This looks good from a cursory review. @peastman do you want to review and merge?

**Comment by peastman** [comment]
I'm not familiar with this part of the code, so I can't comment much on it.  I skimmed through the changes and nothing jumped out that looked obviously wrong.  But since I don't know the algorithm it's supposed to be implementing, I don't have any input on whether it's implemented correctly.

@mpharrigan if you think it looks good, I'm happy to follow your judgement on it.

**Comment by mpharrigan** [comment]
I'm not that familiar either! Gonna go ahead and merge

**Pull request merged**